dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,277 @@
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        cached_dirs = {}
+        cached_ids = {}
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            cached_ids[dir1["id"]] = dir1["title"]
+        return cached_dirs, cached_ids
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            with open(
+                self.gdrive_user_credentials_path, "w"
+            ) as credentials_file:
+                credentials_file.write(
+                    os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA)
+                )
+
+        GoogleAuth.DEFAULT_SETTINGS["client_config_backend"] = "settings"
+        GoogleAuth.DEFAULT_SETTINGS["client_config"] = {
+            "client_id": self.gdrive_client_id,
+            "client_secret": self.gdrive_client_secret,
+            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+            "token_uri": "https://oauth2.googleapis.com/token",
+            "revoke_uri": "https://oauth2.googleapis.com/revoke",
+            "redirect_uri": "",
+        }
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials"] = True
+        GoogleAuth.DEFAULT_SETTINGS["save_credentials_backend"] = "file"
+        GoogleAuth.DEFAULT_SETTINGS[
+            "save_credentials_file"
+        ] = self.gdrive_user_credentials_path
+        GoogleAuth.DEFAULT_SETTINGS["get_refresh_token"] = True
+        GoogleAuth.DEFAULT_SETTINGS["oauth_scope"] = [
+            "https://www.googleapis.com/auth/drive",
+            "https://www.googleapis.com/auth/drive.appdata",
+        ]
+
+        # Pass non existent settings path to force DEFAULT_SETTINGS loading
+        gauth = GoogleAuth(settings_file="")
+        gauth.CommandLineAuth()
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            os.remove(self.gdrive_user_credentials_path)
+
+        gdrive = GoogleDrive(gauth)
+        return gdrive
+
+    def create_drive_item(self, parent_id, title):
+        upload_request = RequestCreateFolder(
+            {"drive": self.drive, "title": title, "parent_id": parent_id}
+        )
+        result = self.execute_request(upload_request)
+        return result
+
+    def get_drive_item(self, name, parents_ids):
+        if not parents_ids:
+            return None
+        query = " or ".join(
+            "'{}' in parents".format(parent_id) for parent_id in parents_ids
+        )
+
+        query += " and trashed=false and title='{}'".format(name)
+
+        list_request = RequestListFile(self.drive, query)
+        item_list = self.execute_request(list_request)
+        return next(iter(item_list), None)
```

Comment:
Should we use `maxItems = 1` here?

Reply:
Makes sense. Resolving of more than 1 item by remote path for further processing is not expected.

Reply:
Resolving this. For the moment it looks as:
```
# Limit found remote items count to 1 in response
        item_list = self.drive.ListFile(
            {"q": query, "maxResults": 1}
        ).GetList()
```